File: flutter/lib/src/ffmpeg_kit_config.dart
/**
 * FFmpegKit Flutter Extended Plugin - A wrapper library for FFmpeg
 * Copyright (C) 2026 Akash Patel
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */

import 'dart:ffi';

import 'callback_manager.dart' as callback_manager;
import 'ffmpeg_kit_extended.dart';
import 'log.dart';
import 'session.dart';
import 'signal.dart';

/// A convenience class for global FFmpegKit configuration.
class FFmpegKitConfig {
  /// Enables global log redirection.
  static void enableRedirection() => FFmpegKitExtended.enableRedirection();

  /// Disables global log redirection.
  static void disableRedirection() => FFmpegKitExtended.disableRedirection();

  /// Sets the global log [level].
  static void setLogLevel(LogLevel level) =>
      FFmpegKitExtended.setLogLevel(level);

  /// Gets the current global log level.
  static LogLevel getLogLevel() => FFmpegKitExtended.getLogLevel();

  /// Sets the directory for font files.
  static void setFontDirectory(String path, {String? mapping}) =>
      FFmpegKitExtended.setFontDirectory(path, mapping: mapping);

  /// Returns the FFmpeg version.
  static String getFFmpegVersion() => FFmpegKitExtended.getFFmpegVersion();

  /// Returns the FFmpegKit version.
  static String getVersion() => FFmpegKitExtended.getVersion();

  /// Returns the FFmpegKit package name.
  static String getPackageName() => FFmpegKitExtended.getPackageName();

  /// Sets the session history [size].
  static void setSessionHistorySize(int size) =>
      FFmpegKitExtended.setSessionHistorySize(size);

  /// Gets the session history size.
  static int getSessionHistorySize() =>
      FFmpegKitExtended.getSessionHistorySize();

  /// Converts log level to string.
  static String? logLevelToString(LogLevel level) =>
      FFmpegKitExtended.logLevelToString(level);

  /// Releases a native handle.
  static void handleRelease(Pointer<Void> handle) =>
      FFmpegKitExtended.handleRelease(handle);

  /// Clears all handled sessions.
  static void clearSessions() => FFmpegKitExtended.clearSessions();

  /// Enables global log callback.
  static void enableLogCallback(
          [callback_manager.FFmpegLogCallback? logCallback]) =>
      FFmpegKitExtended.enableLogCallback(logCallback);

  /// Enables global statistics callback.
  static void enableStatisticsCallback(
          [callback_manager.FFmpegStatisticsCallback? statisticsCallback]) =>
      FFmpegKitExtended.enableStatisticsCallback(statisticsCallback);

  /// Enables global FFmpeg session complete callback.
  static void enableFFmpegSessionCompleteCallback(
          [callback_manager.FFmpegSessionCompleteCallback? completeCallback]) =>
      FFmpegKitExtended.enableFFmpegSessionCompleteCallback(completeCallback);

  /// Enables global FFprobe session complete callback.
  static void enableFFprobeSessionCompleteCallback(
          [callback_manager.FFprobeSessionCompleteCallback?
              completeCallback]) =>
      FFmpegKitExtended.enableFFprobeSessionCompleteCallback(completeCallback);

  /// Enables global FFplay session complete callback.
  static void enableFFplaySessionCompleteCallback(
          [callback_manager.FFplaySessionCompleteCallback? completeCallback]) =>
      FFmpegKitExtended.enableFFplaySessionCompleteCallback(completeCallback);

  /// Enables global MediaInformation session complete callback.
  static void enableMediaInformationSessionCompleteCallback(
          [callback_manager.MediaInformationSessionCompleteCallback?
              completeCallback]) =>
      FFmpegKitExtended.enableMediaInformationSessionCompleteCallback(
          completeCallback);

  /// Registers a new FFmpeg pipe.
  static String? registerNewFFmpegPipe() =>
      FFmpegKitExtended.registerNewFFmpegPipe();

  /// Closes an FFmpeg pipe.
  static void closeFFmpegPipe(String pipePath) =>
      FFmpegKitExtended.closeFFmpegPipe(pipePath);

  /// Sets the list of font directories.
  static void setFontDirectoryList(List<String> fontDirectoryList,
          [Map<String, String>? fontMappings]) =>
      FFmpegKitExtended.setFontDirectoryList(fontDirectoryList, fontMappings);

  /// Returns the build date.
  static String getBuildDate() => FFmpegKitExtended.getBuildDate();

  /// Converts session state to string.
  static String sessionStateToString(SessionState state) =>
      FFmpegKitExtended.sessionStateToString(state);

  /// Parses a command into an argument array.
  static List<String> parseArguments(String command) =>
      FFmpegKitExtended.parseArguments(command);

  /// Converts an argument array to a command string.
  static String argumentsToString(List<String> arguments) =>
      FFmpegKitExtended.argumentsToString(arguments);

  /// Sets an environment variable.
  static void setEnvironmentVariable(String name, String value) =>
      FFmpegKitExtended.setEnvironmentVariable(name, value);

  /// Ignores a specific signal.
  static void ignoreSignal(Signal signal) =>
      FFmpegKitExtended.ignoreSignal(signal);

  /// Returns the number of messages in transmit for a session.
  static int messagesInTransmit(int sessionId) =>
      FFmpegKitExtended.messagesInTransmit(sessionId);
}

File: react-native/src/ffmpeg-kit-config.ts
import {getBackend} from './platform/backend-registry';
import {argumentsToString, parseArguments} from './arguments';
import {LogLevel, SessionState, Signal} from './types';
import {SessionQueueManager} from './session-queue-manager';

const NativeFFmpegKitExtended = getBackend();

/**
 * Global FFmpegKit runtime configuration and utility methods.
 *
 * Settings apply process-wide and affect subsequently executed sessions. Call
 * `FFmpegKitExtended.initialize()` before using native configuration methods.
 */
export class FFmpegKitConfig {
  /**
   * Explicitly enables process-wide native log redirection.
   *
   * Installing an internal completion/log/statistics bridge does not call this
   * method implicitly; the native redirection setting remains user authority.
   */
  static enableRedirection(): void {
    NativeFFmpegKitExtended.enableRedirection();
  }

  /**
   * Explicitly disables process-wide native log redirection.
   *
   * This is authoritative even when a session has an optional log consumer;
   * completion transport remains independent of log/statistics delivery.
   */
  static disableRedirection(): void {
    NativeFFmpegKitExtended.disableRedirection();
  }

  /** Sets the minimum process-wide FFmpeg log level. */
  static setLogLevel(level: LogLevel): void {
    NativeFFmpegKitExtended.setLogLevel(level);
  }

  /** Returns the current process-wide FFmpeg log level. */
  static getLogLevel(): LogLevel {
    return NativeFFmpegKitExtended.getLogLevel() as LogLevel;
  }

  /** Converts a numeric `LogLevel` to the native display name. */
  static logLevelToString(level: LogLevel): string {
    return NativeFFmpegKitExtended.logLevelToString(level);
  }

  /**
   * Registers a directory for fontconfig-based filters such as `drawtext` and
   * `subtitles`.
   *
   * `mapping` optionally maps font family names to filenames in the directory.
   */
  static setFontDirectory(path: string, mapping?: Record<string, string>): void {
    NativeFFmpegKitExtended.setFontDirectory(
      path,
      mapping ? JSON.stringify(mapping) : '',
    );
  }

  /** Sets a native environment variable used by FFmpeg and its libraries. */
  static setEnvironmentVariable(name: string, value: string): void {
    NativeFFmpegKitExtended.setEnvironmentVariable(name, value);
  }

  /** Configures the native runtime to ignore one supported process signal. */
  static ignoreSignal(signal: Signal): void {
    NativeFFmpegKitExtended.ignoreSignal(signal);
  }

  /**
   * Selects the native audio output device by the name returned from
   * `listAudioOutputDevices()`.
   */
  static setAudioOutputDevice(deviceName: string): void {
    NativeFFmpegKitExtended.setAudioOutputDevice(deviceName);
  }

  /**
   * Returns native audio output devices in the wrapper's serialized text form.
   * The exact device names and formatting are platform/backend dependent.
   */
  static listAudioOutputDevices(): string {
    return NativeFFmpegKitExtended.listAudioOutputDevices();
  }

  /** Returns the bundled upstream FFmpeg version. */
  static getFFmpegVersion(): string {
    return NativeFFmpegKitExtended.getFFmpegVersion();
  }

  /** Returns the FFmpegKit Extended wrapper version. */
  static getVersion(): string {
    return NativeFFmpegKitExtended.getVersion();
  }

  /** Returns the selected native package/bundle name. */
  static getPackageName(): string {
    return NativeFFmpegKitExtended.getPackageName();
  }

  /**
   * Sets the maximum number of completed/created sessions retained natively.
   * Reducing the value may discard older history entries.
   */
  static setSessionHistorySize(size: number): void {
    NativeFFmpegKitExtended.setSessionHistorySize(size);
  }

  /** Returns the native session-history capacity. */
  static getSessionHistorySize(): number {
    return NativeFFmpegKitExtended.getSessionHistorySize();
  }

  /**
   * Clears the native session registry and retained session history.
   *
   * Do not call this while sessions are running. Clearing sessions can
   * cancel/invalidate active session handles, and Session getters for cleared
   * IDs may subsequently fail.
   */
  static clearSessions(): void {
    NativeFFmpegKitExtended.clearSessions();
  }

  /**
   * Creates a named FIFO/pipe path for streaming data into or out of FFmpeg.
   * Returns `undefined` when the native platform cannot create the pipe.
   */
  static registerNewFFmpegPipe(): string | undefined {
    return NativeFFmpegKitExtended.registerNewFFmpegPipe() || undefined;
  }

  /** Closes and removes a pipe created by `registerNewFFmpegPipe()`. */
  static closeFFmpegPipe(path: string): void {
    NativeFFmpegKitExtended.closeFFmpegPipe(path);
  }

  /** Returns the enum name for a `SessionState`. */
  static sessionStateToString(state: SessionState): string {
    return SessionState[state] ?? 'Failed';
  }

  /** Splits a command string using the package's platform-neutral parser. */
  static parseArguments(command: string): string[] {
    return parseArguments(command);
  }

  /** Quotes pre-tokenized arguments as one FFmpegKit command string. */
  static argumentsToString(arguments_: readonly string[]): string {
    return argumentsToString(arguments_);
  }

  /**
   * Returns the count of native log/statistics messages still being delivered
   * for a session. This is primarily useful for shutdown diagnostics.
   */
  static messagesInTransmit(sessionId: number): number {
    return NativeFFmpegKitExtended.messagesInTransmit(sessionId);
  }

  /** Returns the JavaScript session queue concurrency limit. */
  static getMaxConcurrentSessions(): number {
    return SessionQueueManager.shared.maxConcurrentSessions;
  }

  /** Sets the JavaScript session queue concurrency limit. */
  static setMaxConcurrentSessions(value: number): void {
    SessionQueueManager.shared.maxConcurrentSessions = value;
  }
}
